### PR TITLE
Parallel Worker Daemons

### PR DESCRIPTION
**Pull request by eskatos**
First step towards the parallel worker API. This PR introduce a public API to submit pieces of work to be executed in parallel from task actions. Only worker daemons are supported.

- [ ] The pull request contains an appropriate level of unit and integration
test coverage to verify the behavior

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java [hunk 1]
```diff
@@ -85,16 +98,78 @@ private GradleException executeActions(TaskInternal task, TaskStateInternal stat
...
+                Throwable actionFailure = null;
+                try {
+                    action.execute(task);
+                } catch (Throwable t) {
+                    actionFailure = t;
+                } finally {
+                    action.contextualise(null);
+                }
+
+                try {
```
Comment (oehme) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/ExecuteActionsTaskExecuter.java]: I feel like "complete a main action and wait for its async work to complete" will be a recurring pattern and should probably be extracted. For instance I can inject the `WorkerExecutor` into a piece of work and schedule more work (or at least I *should* be able to) and my main piece of work should then wait for the additionally scheduled work.

File: subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+        }
+
+        if (failures.size() > 0) {
+            throw workerExecutionException(failures);
+        }
+    }
+
+    private WorkerExecutionException workerExecutionException(List<Throwable> failures) {
+        if (failures.size() == 1) {
+            throw new WorkerExecutionException("There was a failure while executing work items", failures);
```
Comment (oehme) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java]: This message should probably be constructed inside the exception object.

File: subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java [hunk 3]
```diff
@@ -13,40 +13,55 @@
...
-            return workerDaemonWorker.execute(action, spec);
+            return buildOperationExecutor.run(buildOperation, new Transformer<DefaultWorkResult, BuildOperationContext>() {
+                @Override
+                public DefaultWorkResult transform(BuildOperationContext buildOperationContext) {
+                    uses++;
+                    return workerDaemonWorker.execute(action, spec);
+                }
+            });
         } finally {
             workerLease.operationFinish();
```
Comment (oehme) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java]: Shouldn't the buildOperationExecutor already take care of limiting work in progress? Why do we use mechanisms here? As far as I understood the explicit worker lease mechanism was a low-level concept to be used only when you can't conveniently package the work to be performed in a callback.

Comment (eskatos) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java]: This `uses` fields account for the number of times a worker daemon has been used. Its not for limiting work in progress. It is then used to take expiration decisions.

**Comment by eskatos** [comment]
@oehme 
> What's not quite clear to me is what level of abstraction our compiler daemons and dependency transformations should use. Should they use the new WorkerExecutor or something below?

I'd say that they should use the `WorkerExecutor` public API. So that they get the possibility to fork or not, use reusable worker daemons or not etc.. Plus it's a good test drive for the new API, dogfooding, if the API is not sufficient it probably means that we need to make it better.

**Comment by ghale** [comment]
In order for the compilers to use the new API, we would need to support the `Callable` use case.  This is certainly something we should do, and should not be difficult, but I question whether it's necessary for this story.  Personally, I'd rather not heap any more onto this and would prefer to move that out to another story.

**Comment by oehme** [comment]
I definitely didn't want to suggest we do this as part of this story. I just want to understand our general direction. There are now a lot of internal APIs for tracking/scheduling work and it might become confusing which one to use in what scenario. If we can unify on one in the mid term, that would be great.

File: subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java (hunk 3 above)
Comment (oehme) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java]: I'm referring to the fact that we are using the `workerLease` API here, even though we are already using the buildOperationExecutor (which should get a worker lease under the hood anyway).

Comment (ghale) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java]: We _really_ need better terminology for this stuff.  We currently have 3 distinct things that we refer to as "build operations" and I think this creates a lot of confusion.  There are "build operations" that are started from `BuildOperationExecutor` which handles firing events to be consumed by build scans, et al.  There are "build operations" that are started from `BuildOperationWorkerRegistry` for limiting parallel work based on `max-workers`.  And there are "build operations" that are created by `BuildOperationProcessor` which is also used to limit parallel work.  `BuildOperationWorkerRegistry` is used when `BuildOperationProcessor` cannot be used to limit parallel work, which is why we are using it here.  `BuildOperationExecutor` only handles the firing of build operation events and progress logging.

Comment (eskatos) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java]: Ah right!

Even though `BuildOperationWorkerRegistry` and `BuildOperationExecutor` manage leases cooperatively, they still have a different "build operation" model. This PR introduce an unwanted "build operation" to bridge the two models.

`BuildOperationWorkerRegistry` and `BuildOperationExecutor` need some more love.
Good catch!

Comment (eskatos) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java]: Mmm, scratch my comment above ☝️.

`BuildOperationWorkerRegistry` and `BuildOperationProcessor` execute build operations in parallel. `BuildOperationExecutor` does not, it has nothing to do with worker leases.

Comment (oehme) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonClient.java]: Thanks @ghale, I confused the two as you correctly pointed out. We need better names for this stuff.

File: subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    void registerAsyncWork(final RunnableWorkFuture result) {
+        asyncWorkTracker.registerWork(buildOperationExecutor.getCurrentOperation(), new AsyncWorkCompletion() {
+            @Override
+            public void waitForCompletion() {
+                try {
+                    result.getIfNotHandled();
+                } catch (InterruptedException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
+                } catch (ExecutionException e) {
+                    throw UncheckedException.throwAsUncheckedException(e);
```
Comment (adammurdoch) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java]: should probably be `e.getCause()`

File: subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    @Contextual
+    private static class WorkExecutionException extends ExecutionException {
+        public WorkExecutionException(String description) {
+            this(description, null);
+        }
+
+        public WorkExecutionException(String description, Throwable cause) {
+            super("A failure occurred while executing " + description + ": " + cause.getMessage(), cause);
```
Comment (adammurdoch) [inline subprojects/workers/src/main/java/org/gradle/workers/internal/DefaultWorkerExecutor.java]: We should avoid this pattern of including the cause's message in the contextual exception's message

File: subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelIntegrationTest.groovy [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+package org.gradle.workers.internal
+
+import org.gradle.test.fixtures.server.http.BlockingHttpServer
+import org.junit.Rule
+import spock.lang.Timeout
+import spock.lang.Unroll
+
+@Timeout(60)
+class WorkerExecutorParallelIntegrationTest extends AbstractWorkerExecutorIntegrationTest {
```
Comment (adammurdoch) [inline subprojects/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParallelIntegrationTest.groovy]: good test coverage.

We should add tests that show:

- A task that depends on a task that runs a bunch of actions does not start until the actions have been completed.
- `--max-workers` is honored when running actions in parallel.
- That a TAPI client receives build operation events for each of the task actions and the work actions.

**Pull request merged**